rust/examples/battery_service.rs

```diff
@@ -0,0 +1,188 @@
+    let address = Address::from_be_hex("F0:F1:F2:F3:F4:F5", AddressType::RandomDeviceAddress)
+        .map_err(|e| anyhow!(e))?;
+    let device =
+        Device::with_hci("Bumble", address, transport.source()?, transport.sink()?).await?;
+    device.power_on().await?;
+
+    match cli.subcommand {
+        Subcommand::Client { target_addr } => {
+            let target = Address::from_be_hex(&target_addr, AddressType::RandomDeviceAddress)
+                .map_err(|e| anyhow!("{:?}", e))?;
+            run_client(device, target).await
+        }
+        Subcommand::Server => run_server(device).await,
+    }?;
+
+    // wait until user kills the process
+    tokio::signal::ctrl_c().await?;
+    Ok(())
+}
+
+async fn run_client(device: Device, target_addr: Address) -> anyhow::Result<()> {
+    info!("Connecting");
+    let conn = device.connect(&target_addr).await?;
+    let mut peer = Peer::new(conn).await?;
+    info!("Discovering");
+    peer.discover_services().await?;
+    peer.discover_characteristics().await?;
+
+    let battery_service = peer
+        .create_service_proxy::<BatteryServiceProxy>()?
+        .ok_or(anyhow::anyhow!("No battery service found"))?;
+
+    info!("Getting characteristic");
+    let mut battery_level_char = battery_service
+        .battery_level()?
+        .ok_or(anyhow::anyhow!("No battery level characteristic"))?;
+    info!("Reading");
+    info!(
+        "{} {}",
+        "Initial Battery Level:".green(),
+        battery_level_char.read_value().await?
+    );
+    info!("Subscribing");
+    battery_level_char
+        .subscribe(|value| {
+            info!("{} {:?}", "Battery level update:".green(), value);
+        })
+        .await?;
+
+    Ok(())
+}
+
+async fn run_server(mut device: Device) -> anyhow::Result<()> {
+    let uuid = services::BATTERY.uuid();
+    let battery_level_uuid = Uuid16::from(0x2A19).into();
+    let battery_level = Characteristic::new(
+        battery_level_uuid,
+        CharacteristicProperty::Read | CharacteristicProperty::Notify,
+        AttributePermission::Readable.into(),
+        CharacteristicValueHandler::new(Box::new(BatteryRead), Box::new(NoOpWrite)),
+    );
+    let service = Service::new(uuid.into(), vec![battery_level]);
+    let service_handle = device.add_service(&service)?;
+
+    let mut builder = AdvertisementDataBuilder::new();
+    builder.append(CommonDataType::CompleteLocalName, "Bumble Battery")?;
+    builder.append(
+        CommonDataType::IncompleteListOf16BitServiceClassUuids,
+        &uuid,
+    )?;
+    builder.append(
+        CommonDataType::Appearance,
+        // computer (0x02) - laptop (0x03)
+        [0x02, 0x03].as_slice(),
+    )?;
+    device.set_advertising_data(builder)?;
+    device.power_on().await?;
+    device.start_advertising(true).await?;
+
+    let char_handle = service_handle
+        .characteristic_handle(battery_level_uuid)
+        .expect("Battery level should be present");
+
+    loop {
```

Comment:
It could be done with a cancellation token and a `select!` in the loop, but would it really change anything? Either way they both just die when ^C'd. There's no meaningful teardown to be done.

Reply:
In the past (not sure if the behavior is still the same), if you force-quit without calling "stop advertising", the dongle will continue advertising.

That said, in the past, there were issues with the `asyncio` library inadvertently setting up a `KeyboardInterrupt` signal handler when it imported the `signal` library, so even if you gracefully await ctrl-c, it exits before "stop advertising" gets called.

Though at the end of the day, this is just an example. So if you want to keep it simple, I can vibe with that.

Reply:
Hm, I suppose it would be better for readers of the example to be able to see best practices in all of them. I'll add in ctrl-c handling.